cpp/src/io/parquet/chunk_dict.cu

```diff
@@ -102,82 +112,74 @@ __global__ void __launch_bounds__(block_size, 1)
 
-  if (t == 0) {
-    // Find the bounds of values in leaf column to be inserted into the map for current chunk
-    auto cudf_col      = *(col->parent_column);
-    s_start_value_idx  = row_to_value_idx(start_row, cudf_col);
-    auto end_value_idx = row_to_value_idx(end_row, cudf_col);
-    s_num_values       = end_value_idx - s_start_value_idx;
-  }
-  __syncthreads();
+  // Find the bounds of values in leaf column to be inserted into the map for current chunk
+  auto const cudf_col          = *(col->parent_column);
+  auto const s_start_value_idx = row_to_value_idx(start_row, cudf_col);
+  auto const end_value_idx     = row_to_value_idx(end_row, cudf_col);
 
   column_device_view const& data_col = *col->leaf_column;
-  using block_reduce                 = cub::BlockReduce<size_type, block_size>;
-  __shared__ typename block_reduce::TempStorage reduce_storage;
 
   // Make a view of the hash map
   auto hash_map_mutable = map_type::device_mutable_view(
     chunk->dict_map_slots, chunk->dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
-  auto hash_map = map_type::device_view(
-    chunk->dict_map_slots, chunk->dict_map_size, KEY_SENTINEL, VALUE_SENTINEL);
 
   __shared__ int total_num_dict_entries;
-  for (size_type i = 0; i < s_num_values; i += block_size) {
-    // add the value to hash map
-    size_type val_idx = i + t + s_start_value_idx;
-    bool is_valid =
-      (i + t < s_num_values && val_idx < data_col.size()) and data_col.is_valid(val_idx);
+  auto val_idx = s_start_value_idx + t;
+  while (block.any(val_idx < end_value_idx)) {
+    auto const is_valid = val_idx < data_col.size() and data_col.is_valid(val_idx);
 
     // insert element at val_idx to hash map and count successful insertions
     size_type is_unique      = 0;
     size_type uniq_elem_size = 0;
     if (is_valid) {
-      auto found_slot = type_dispatcher(data_col.type(), map_find_fn{hash_map}, data_col, val_idx);
-      if (found_slot == hash_map.end()) {
-        is_unique =
-          type_dispatcher(data_col.type(), map_insert_fn{hash_map_mutable}, data_col, val_idx);
-        uniq_elem_size = [&]() -> size_type {
-          if (not is_unique) { return 0; }
-          switch (col->physical_type) {
-            case Type::INT32: return 4;
-            case Type::INT64: return 8;
-            case Type::INT96: return 12;
-            case Type::FLOAT: return 4;
-            case Type::DOUBLE: return 8;
-            case Type::BYTE_ARRAY:
-              if (data_col.type().id() == type_id::STRING) {
-                // Strings are stored as 4 byte length + string bytes
-                return 4 + data_col.element<string_view>(val_idx).size_bytes();
-              }
-            case Type::FIXED_LEN_BYTE_ARRAY:
-              if (data_col.type().id() == type_id::DECIMAL128) { return sizeof(__int128_t); }
-            default: CUDF_UNREACHABLE("Unsupported type for dictionary encoding");
-          }
-        }();
-      }
+      is_unique =
```

Comment:
This was a small optimization effort. I found that find and insert was faster than trying to insert first. I think this was because `find` didn't use a CAS. And there's a much higher likelihood of having more values than the max dictionary size. My memory is a bit hazy because this is a year old.

Reply:
Ah, I can understand then. Just compared the old `find` + `insert` again `insert` and found no performance difference. I will keep the change since it simplifies the code a bit.

Reply:
By varying data sizes and cardinalities, I see `populate_chunk_hash_maps_kernel` may take from 5 ms to 20 ms. Then I fix the data size, the performance difference is trivial and I found it's hard to tell whether the difference is from the implementation:

- cardinality  = 1000:
  - `find`+`insert`: 6.1 ms
  - `find`: 5.9 ms
- cardinality = 10'000'000:
  - `find`+`insert`: 8.9 ms
  - `find`: 9.3 ms

Reply:
Did you mean find+insert vs **insert**?
Alright, then it's fine. I'm sure I saw a perf improvement but maybe it was in a previous version of this code and it's not as useful now.

Reply:
`insert` it is. Typo corrected.